### PR TITLE
[TRTLLM-15316][fix] SM107 FP8 GEMM routing and runtime guards (follow-up to #17485)

### PR DESCRIPTION
**Pull request by farazkh80**
## Description

Follow-up to #17485 ("[TRTLLM-15316][feat] sm107 gemm + quant"). That PR shipped the C++-only
slice; this one lands the Python wiring that goes with it: SM107 FP8 GEMM routing correctness, a
couple of `arch == 100 || arch == 103` sites that should use the existing
`isSM100Family`/`is_sm_100f` helpers, two arch-neutral bugfixes that ride along, and a couple of
small SM107-gated runtime guards.

A separate CuTe DSL PR is coming next that adds the FP8 CuTe DSL GEMM/BMM runners for SM107 — none
of that is included here.

Credits: items 2 and 5 below were originally authored by Yifei Zhang and item 4 by Bowen Fu
(`Co-authored-by` trailers on those commits).

## Scope

1. **FP8 block-scale GEMM routing on SM107** (`tensorrt_llm/_torch/modules/linear.py`,
   `tests/unittest/_torch/moe/quantize_utils.py`) — `cute_dsl_fp8_gemm_blackwell` runs only on
   sm100/103. When `use_cute_dsl_blockscaling_mm` or `disable_deep_gemm` is set on another `sm_100f`
   GPU (e.g. sm107), route to `fp8_block_scaling_gemm` instead: its sm100-family branch is the
   trtllm-gen kernel, which has sm107 cubins and consumes the same raw fp32 scales as the CuTe DSL
   path — so `disable_deep_gemm` keeps meaning "not DeepGEMM" (Qwen3 / EXAONE4 set it for DeepGEMM
   correctness issues). A one-time warning is logged when `use_cute_dsl_blockscaling_mm` falls
   through, since the requested kernel is genuinely unavailable there. `transform_weights()` is
   unchanged: both non-DeepGEMM paths consume raw scales, so main's existing condition already
   matches.
2. **`isSM100Family`/`is_sm_100f` consistency** (`cpp/tensorrt_llm/kernels/weightOnlyBatchedGemv/kernelLauncher.h`,
   `tensorrt_llm/_torch/attention/backends/sparse/rocket/kernels.py`, `tensorrt_llm/quantization/functional.py`) —
   residue of an already-merged fix (the cutlass_preprocessors/fpA_intB C++ hunks landed in #17485;
   this is the remaining Python + one C++ dispatch site).
3. **`enable_attention_dp` propagation into overridden-TP sub-`Mapping`** (`gated_mlp.py`, `mlp.py`) —
   arch-neutral bugfix, nvbugs/6568929.
4. **Safe SM107 MoE fallback tactic** (`custom_ops/torch_custom_ops.py` `MoERunner._resolve_fallback_tactic`) —
   gated to sm107 + bf16 x bf16; falls back to the runner's shape-safe final tactic when the autotuner
   has no cached tactic (`tactic == -1`).
5. **CUDA Tile RMSNorm SM107 guard** (`custom_ops/cuda_tile_custom_ops.py`) — no CUDA Tile kernels are
   built for sm107; raise `RuntimeError` instead of silently running undefined behavior, and skip the
   corresponding unit tests on sm107.

## Blast radius

| File | Gate | Non-SM107 behavior |
|---|---|---|
| `modules/linear.py` (`apply`) | inside the existing `use_cute_dsl_blockscaling_mm or disable_deep_gemm` branch: `sm_version in (100, 103)` → CuTe DSL, else → trtllm-gen `fp8_block_scaling_gemm` | SM100/SM103 unchanged (still take the CuTe DSL path); the DeepGEMM `fp8_swap_ab_gemm` branch and the SM90/SM120 paths are untouched |
| `kernelLauncher.h` | `isSM100Family(arch)` (100 <= arch < 110) | Identical dispatch for arch 100/103 (only prior members of the set); SM90 branch untouched |
| `attention/backends/sparse/rocket/kernels.py` (`triton_bmm`) | `get_sm_version() == 107` picks `num_warps=8` | All other archs keep `num_warps=4` (unchanged) |
| `quantization/functional.py` (`preprocess_weights_for_mixed_gemm`) | `is_sm_100f(sm_)` (100 <= sm < 110) | Identical result for sm 100/103; other archs unaffected |
| `gated_mlp.py`, `mlp.py` | none — arch-neutral bugfix | All archs: overridden-TP sub-`Mapping` now correctly inherits `enable_attention_dp` from the parent mapping instead of resetting it |
| `custom_ops/torch_custom_ops.py` (`MoERunner`) | `get_sm_version() == 107 and x_dtype == weight_dtype == bfloat16` | All other arch/dtype combinations: `_resolve_fallback_tactic` is a no-op passthrough |
| `custom_ops/cuda_tile_custom_ops.py` | `get_sm_version() in [107]` | All other archs: unchanged (no guard triggers) |

## Test coverage

SM107 hardware doesn't exist to test on directly, so verification here is: (a) the new/changed unit
tests, and (b) a B300 (sm100/103, Blackwell) regression pass confirming every touched path is an
exact no-op off SM107, run against a local build of this branch (`--cuda_architectures=103-real`):

- `tests/unittest/_torch/moe/test_moe_runner.py` (new) — 6/6 passed. Covers
  `MoERunner._resolve_fallback_tactic`: sm107+bf16 fallback, all other arch/dtype combinations as
  no-ops, tuned-tactic passthrough.
- `tests/unittest/_torch/thop/parallel_hw_agnostic/test_cuda_tile_custom_ops.py` — 229/229 passed on
  B300. The `@skip_rubin` markers only skip on sm107, so the existing 227 are a full regression run for
  the guard's else-branch; the 2 new `*_rejects_sm107` tests monkeypatch `get_sm_version()` to 107 and
  assert both ops raise, covering the guard itself without SM107 hardware.
- `tests/unittest/_torch/modules/test_linear_uneven_tp.py::TestFP8BlockScalesMLP` — 3/3 passed on
  B300. Exercises both the `linear.py` FP8 routing fix and the `enable_attention_dp` sub-`Mapping`
  fix end to end.
- The sm107 fallback kernel itself (`fp8_quantize_1x128` + `fp8_block_scaling_gemm`, raw scales) is
  main's existing `test_fp8_block_scale_gemm` path; its body passes on B300 for
  `(m,k,n) ∈ {(128,7168,2112), (7,1536,24576), (4096,2048,7168), (64,1024,1024)}`
  (`calc_diff < 1e-3`, `assert_close(atol=rtol=1e-3)`). The test's own sm list already includes 107.
- `tests/unittest/_torch/thop/parallel/test_weight_only_quant_gemm.py` +
  `test_weight_only_quant_linear.py` — 68/68 passed on B300. Exercises the `kernelLauncher.h`
  `isSM100Family` dispatch.
- `tests/unittest/_torch/moe/test_moe_module.py -k FP8BlockScales` — 17 passed / 78 skipped
  (hardware-gated variants) on B300. Exercises `RefMLPFusedMoE.post_load_weights()`.

Not independently verifiable pre-merge: the sm107-only branches themselves (routing to
`fp8_swap_ab_gemm`, the `num_warps=8` triton config, the MoE fallback tactic, the CUDA Tile
`RuntimeError`, and the relaxed sm107 accuracy tolerance in `quantize_utils.py`) — there's no SM107
hardware to run them on. These are covered by code review and by the mock-based
`test_moe_runner.py` cases that force `get_sm_version() == 107`.

## Checklist

- [x] `disclosure_gate.sh` exit 0; every added comment read manually
- [x] Pre-commit clean (isort, yapf, autoflake, clang-format, codespell, ruff, ruff-format, ruff-legacy)
- [x] Commits signed off (`-s`), committer is the noreply identity, original authors preserved via `--author=`
- [x] Local build + targeted unit tests green on B300 (see Test coverage)

Note: CI (`/bot run`) is triggered manually by Faraz, not automated for this PR.

[x] Please check this after reviewing the above items as appropriate for this PR.

## Dev Engineer Review

- Routes SM107 FP8 block-scaled GEMM requests to `fp8_block_scaling_gemm` when required.
- Preserves CuTe DSL routing for SM100 and SM103.
- Adds a one-time warning when CuTe DSL is unavailable.
- Uses `isSM100Family` and `is_sm_100f` for architecture checks.
- Configures eight Triton warps for SM107.
- Preserves `enable_attention_dp` in overridden tensor-parallel mappings.
- Adds an SM107 BF16 MoE tactic fallback.
- Rejects unsupported SM107 CUDA Tile RMSNorm operations with `RuntimeError`.
- Adds targeted coverage for the MoE fallback and RMSNorm rejection.
- Restores `transform_weights()` behavior.
- B300 validation passed for the applicable FP8 block-scale and fallback GEMM paths.
- SM107 hardware validation was not available.
- Review focus: verify CI failures in runs 58635, 58674, and 58698, and investigate the unstable result in run 58705.

## QA Engineer Review

Modified test functions include:

- MoE fallback tactic tests for SM107 BF16 inputs.
- Tests that preserve `-1` for unsupported architectures and dtype combinations.
- Tests that preserve explicitly tuned tactics.
- CUDA Tile RMSNorm rejection tests for both RMSNorm operations.
- Existing CUDA Tile RMSNorm tests updated with Rubin skips.
- MXFP4 and MXFP8 reference accuracy checks updated for SM107.

The summary does not identify matching entries in `tests/integration/test_lists/`, `test-db/`, or `qa/`. Coverage registration requires follow-up.

**Verdict: needs follow-up.**

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 1]
```diff
@@ -2832,7 +2832,7 @@ def mla_rope_generation(
             self.num_kv_heads,
             self.head_dim,
             metadata.kv_cache_manager.tokens_per_block,
-            metadata.max_seq_len,  # attention_window_size
+            int(metadata.max_seq_len),  # attention_window_size
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: Seems unnecessary

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py [hunk 2]
```diff
@@ -1385,7 +1385,7 @@ def _call_paged_mqa_logits(
                 context_lens,
                 block_table,
                 scheduler_metadata_buffer,
-                max_seq_len,
+                int(max_seq_len),
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py]: Seems unnecessary

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py [hunk 3]
```diff
@@ -1394,7 +1394,7 @@ def _call_paged_mqa_logits(
             context_lens,
             block_table,
             scheduler_metadata_buffer,
-            max_seq_len,
+            int(max_seq_len),
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py]: Seems unnecessary

**Review by yuxianq — APPROVED** [review]
Attention part LGTM

File: tensorrt_llm/_torch/attention_backend/trtllm.py (hunk 1 above)
Comment (farazkh80) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: Good catch, dropped. Checked: `metadata.kv_cache_manager.max_seq_len` is already passed uncast to the same attention_window_size arg at lines 2617/2663/2703 in this file, so the cast was redundant. Reverted in `a44d7383f7`.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py (hunk 2 above)
Comment (farazkh80) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py]: Same as the trtllm.py comment — this max_seq_len value is also passed uncast at the cute_dsl_*_paged_mqa_logits call sites a few lines down, so the cast was redundant. Reverted in `a44d7383f7`.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py (hunk 3 above)
Comment (farazkh80) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/indexer.py]: Same fix, reverted in `a44d7383f7`.

File: tests/unittest/_torch/thop/parallel_hw_agnostic/test_cuda_tile_custom_ops.py [hunk 4]
```diff
@@ -98,6 +99,7 @@ def reference_rms_norm(
         return hidden_states
 
 
+@skip_rubin
```
[the remark this answers is not in the document]
Comment (farazkh80) [reply tests/unittest/_torch/thop/parallel_hw_agnostic/test_cuda_tile_custom_ops.py:142]: Added `test_cuda_tile_rms_norm_rejects_sm107` and `test_cuda_tile_rms_norm_fuse_residual_rejects_sm107` in `550754099e` — they monkeypatch `get_sm_version()` to 107 and assert both ops raise, so the rejection path runs on any GPU without needing real SM107 hardware. Both pass on B300.

**Review by sylvesterkaczmarek — COMMENTED** [review]
One edge case on SM107: this makes `disable_deep_gemm=True` ineffective. `uses_cute_dsl_path` is forced false for SM107, so the `is_sm_100f()` branch falls through to `fp8_swap_ab_gemm`, which is the DeepGEMM path. If cute-dsl is unavailable on SM107, could this reject that configuration or use another fallback instead of silently ignoring the explicit disable flag?

**Review by BowenFu — APPROVED** [review]

**Review by xxi-nv — APPROVED** [review]

**Review by QiJune — APPROVED** [review]
LGTM

**Comment by farazkh80** [comment]
Thanks, @sylvesterkaczmarek — good catch. Fixed in `c7a298855c`: on sm_100f GPUs other than sm100/103, `use_cute_dsl_blockscaling_mm` / `disable_deep_gemm` now route to `fp8_block_scaling_gemm` (trtllm-gen, has sm107 cubins, same raw fp32 scales as the CuTe DSL path), so `disable_deep_gemm` still means "not DeepGEMM". `use_cute_dsl_blockscaling_mm` logs a one-time warning there. `transform_weights()` is back to main's version.

Verified on B300: `TestFP8BlockScalesMLP` 3/3, and main's `test_fp8_block_scale_gemm` body passes on the fallback op.

File: tests/unittest/_torch/thop/parallel_hw_agnostic/test_cuda_tile_custom_ops.py [hunk 5]
```diff
@@ -98,6 +99,45 @@ def reference_rms_norm(
         return hidden_states
 
 
+def test_cuda_tile_rms_norm_rejects_sm107(monkeypatch):
```
[the remark this answers is not in the document]
Comment (farazkh80) [reply tests/unittest/_torch/thop/parallel_hw_agnostic/test_cuda_tile_custom_ops.py]: Done in `19dd53385f` — both tests now take `monkeypatch: pytest.MonkeyPatch` and return `None`.

**Review by sylvesterkaczmarek — COMMENTED** [review]
Re-reviewed the SM107 `disable_deep_gemm` edge case. On non-100/103 SM100f devices the opt-out now routes through `fp8_block_scaling_gemm` rather than the DeepGEMM path, while explicit CuTe DSL requests warn about the fallback. My previous blocker is resolved.
